File: backend/src/routes/progress.js
import express from 'express';
import mongoose from 'mongoose';
const router = express.Router();
import { auth } from '../middleware/auth.js';
import ProgressLog from '../models/ProgressLog.js';
import Task from '../models/Task.js';
import Notification from '../models/Notification.js';
import Activity from '../models/Activity.js';
import User from '../models/User.js';

router.get('/task/:taskId', auth, async (req, res) => {
  try {
    const logs = await ProgressLog.find({ taskId: req.params.taskId })
      .populate('userId', 'name email avatarUrl')
      .sort({ createdAt: -1 });

    res.json(logs.map(log => ({
      id: log._id.toString(),
      taskId: log.taskId.toString(),
      userId: log.userId._id.toString(),
      date: log.date.toISOString(),
      progressText: log.progressText,
      percentageComplete: log.percentageComplete,
      hoursSpent: log.hoursSpent,
      attachments: log.attachments,
      links: log.links,
      feedbackStatus: log.feedbackStatus,
      adminFeedback: log.adminFeedback,
      createdAt: log.createdAt.toISOString(),
    })));
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post('/', auth, async (req, res) => {
  try {
    const logData = {
      ...req.body,
      userId: req.user._id,
      taskId: mongoose.Types.ObjectId.isValid(req.body.taskId) 
        ? new mongoose.Types.ObjectId(req.body.taskId) 
        : req.body.taskId,
      date: req.body.date ? new Date(req.body.date) : new Date(),
    };

    const log = new ProgressLog(logData);
    await log.save();

    await Task.findByIdAndUpdate(req.body.taskId, { updatedAt: new Date() });

    const task = await Task.findById(req.body.taskId).populate('createdBy');
    if (task && task.createdBy._id.toString() !== req.user._id.toString()) {
      const notification = new Notification({
        userId: task.createdBy._id,
        type: 'progress_submitted',
        message: `${req.user.name} submitted progress for '${task.title}'`,
        payload: { taskId: task._id.toString(), progressId: log._id.toString() },
      });
      await notification.save();
    }

    res.status(201).json({
      id: log._id.toString(),
      taskId: log.taskId.toString(),
      userId: log.userId.toString(),
      date: log.date.toISOString(),
      progressText: log.progressText,
      percentageComplete: log.percentageComplete,
      hoursSpent: log.hoursSpent,
      attachments: log.attachments,
      links: log.links,
      feedbackStatus: log.feedbackStatus,
      createdAt: log.createdAt.toISOString(),
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post('/:id/approve', auth, async (req, res) => {
  try {
    if (req.user.role !== 'admin') {
      return res.status(403).json({ error: 'Admin access required' });
    }

    const { feedback } = req.body;
    const log = await ProgressLog.findByIdAndUpdate(
      req.params.id,
      {
        feedbackStatus: 'approved',
        adminFeedback: feedback || '',
      },
      { new: true }
    ).populate('userId', 'name');

    if (!log) {
      return res.status(404).json({ error: 'Progress log not found' });
    }

    const notification = new Notification({
      userId: log.userId._id,
      type: 'progress_approved',
      message: `Your progress was approved`,
      payload: { taskId: log.taskId.toString(), progressId: log._id.toString() },
    });
    await notification.save();

    const activity = new Activity({
      userId: req.user._id,
      userName: req.user.name,
      action: 'approved progress',
      targetType: 'progress',
      targetId: log._id.toString(),
    });
    await activity.save();

    res.json({
      id: log._id.toString(),
      feedbackStatus: log.feedbackStatus,
      adminFeedback: log.adminFeedback,
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

router.post('/:id/reject', auth, async (req, res) => {
  try {
    if (req.user.role !== 'admin') {
      return res.status(403).json({ error: 'Admin access required' });
    }

    const { feedback } = req.body;
    const log = await ProgressLog.findByIdAndUpdate(
      req.params.id,
      {
        feedbackStatus: 'rejected',
        adminFeedback: feedback || '',
      },
      { new: true }
    ).populate('userId', 'name');

    if (!log) {
      return res.status(404).json({ error: 'Progress log not found' });
    }

    const notification = new Notification({
      userId: log.userId._id,
      type: 'progress_rejected',
      message: `Your progress was rejected`,
      payload: { taskId: log.taskId.toString(), progressId: log._id.toString() },
    });
    await notification.save();

    res.json({
      id: log._id.toString(),
      feedbackStatus: log.feedbackStatus,
      adminFeedback: log.adminFeedback,
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

File: backend/app/routes/progress.py
from datetime import date, datetime
from fastapi import APIRouter, Depends, HTTPException, status
from app.auth import get_current_user, require_admin, transform_record
from app.database import get_admin_client
from app.schemas.progress_log import ProgressLogCreate
from app.schemas.feedback import FeedbackRequest

router = APIRouter()


@router.get("/task/{task_id}")
async def get_progress_logs(
    task_id: str, current_user: dict = Depends(get_current_user)
):
    supabase = get_admin_client()
    response = (
        supabase.table("progress_logs")
        .select("*")
        .eq("task_id", task_id)
        .order("created_at", desc=True)
        .execute()
    )
    return [transform_record(log) for log in response.data]


@router.post("")
async def create_progress_log(
    body: ProgressLogCreate, current_user: dict = Depends(get_current_user)
):
    supabase = get_admin_client()
    log_data = {
        "task_id": body.taskId,
        "user_id": current_user["id"],
        "progress_text": body.progressText,
        "percentage_complete": body.percentageComplete,
        "hours_spent": body.hoursSpent,
        "attachments": body.attachments or [],
        "links": body.links or [],
        "date": body.date or date.today().isoformat(),
        "feedback_status": "pending",
    }

    response = supabase.table("progress_logs").insert(log_data).execute()
    log = response.data[0]

    supabase.table("tasks").update({"updated_at": datetime.utcnow().isoformat()}).eq(
        "id", body.taskId
    ).execute()

    task_response = (
        supabase.table("tasks")
        .select("created_by, title")
        .eq("id", body.taskId)
        .execute()
    )
    if task_response.data:
        task = task_response.data[0]
        if task["created_by"] != current_user["id"]:
            supabase.table("notifications").insert(
                {
                    "user_id": task["created_by"],
                    "type": "progress_submitted",
                    "message": f"{current_user['name']} submitted progress for '{task['title']}'",
                    "payload": {"taskId": body.taskId, "progressId": log["id"]},
                }
            ).execute()

    return transform_record(log)


@router.post("/{log_id}/approve")
async def approve_progress(
    log_id: str, body: FeedbackRequest, current_user: dict = Depends(require_admin)
):
    supabase = get_admin_client()
    response = (
        supabase.table("progress_logs")
        .update(
            {
                "feedback_status": "approved",
                "admin_feedback": body.feedback or "",
            }
        )
        .eq("id", log_id)
        .execute()
    )

    if not response.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Progress log not found"
        )

    log = response.data[0]

    supabase.table("notifications").insert(
        {
            "user_id": log["user_id"],
            "type": "progress_approved",
            "message": "Your progress was approved",
            "payload": {"taskId": log["task_id"], "progressId": log["id"]},
        }
    ).execute()

    supabase.table("activities").insert(
        {
            "user_id": current_user["id"],
            "user_name": current_user["name"],
            "action": "approved progress",
            "target_type": "progress",
            "target_id": log["id"],
        }
    ).execute()

    return {
        "id": log["id"],
        "feedbackStatus": log["feedback_status"],
        "adminFeedback": log["admin_feedback"],
    }


@router.post("/{log_id}/reject")
async def reject_progress(
    log_id: str, body: FeedbackRequest, current_user: dict = Depends(require_admin)
):
    supabase = get_admin_client()
    response = (
        supabase.table("progress_logs")
        .update(
            {
                "feedback_status": "rejected",
                "admin_feedback": body.feedback or "",
            }
        )
        .eq("id", log_id)
        .execute()
    )

    if not response.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Progress log not found"
        )

    log = response.data[0]

    supabase.table("notifications").insert(
        {
            "user_id": log["user_id"],
            "type": "progress_rejected",
            "message": "Your progress was rejected",
            "payload": {"taskId": log["task_id"], "progressId": log["id"]},
        }
    ).execute()

    return {
        "id": log["id"],
        "feedbackStatus": log["feedback_status"],
        "adminFeedback": log["admin_feedback"],
    }
